Bluetooth: Update indication callback parameters

JordanYates opened the pull request:
Implementing the RFC described in #29357

Closes #29357

Commit:
Bluetooth: indication callback signature update

Update the signature of the `bt_gatt_indicate_func_t` callback type by
replacing the attr pointer with a pointer to the
`bt_gatt_indicate_params` struct that was used to start the indication.

This allows the callback to free the `bt_gatt_indicate_params` instance
if it was allocated from storage, while still allowing the
`bt_gatt_attr` value to be accessed through `params->attr`.

Allocating the `bt_gatt_indicate_params` instance from storage is
desirable as multiple indications can be queued, however each instance
must be valid until the callback is run.

Implements API update from #29357

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

Commit:
Bluetooth: add destroy callback to indication

Adds a `destroy` callback to the `struct bt_gatt_indicate_params` which
is used to signify to the application that the indication operation has
completed and the struct instance can be freed/destroyed.

This is required as the number of indication value callbacks that will
be triggered is not known by the caller when the `conn` parameter is
`NULL`.

Tracking when this callback should be run is mananged by a private
reference counter inside the struct.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

carlescufi commented:
> Implementing the RFC described in #29357

Excellent, thank you! Could you please send an email to the mailing list as well, as described in the process?

carlescufi reviewed (approved):
Approved in API meeting

JordanYates commented:
Rebased, `usage_count` renamed to `ref`, documentation updated.

joerchan reviewed (approved):
Thank you.

jhedberg commented on include/bluetooth/gatt.h:
```diff
@@ -1018,6 +1021,8 @@ struct bt_gatt_indicate_params {
 	/** Indicate Value callback */
 	bt_gatt_indicate_func_t func;
+	/* Host populated counter, this struct can be freed when equal to 0 */
+	int usage_count;
```
Are the semantics the same here though? I.e. is this strictly a count for how many pointer variables are referring to the same parameters object? If so, I'd expect to see a clearing (setting to NULL or going out of scope) of a pointer variable every time the reference is decremented, and copying of a pointer variable (to another pointer variable) whenever the reference count is incremented. Looking at the changes that doesn't seem to be completely true here?

joerchan replied:
GATT creates X ATT requests, that each has a pointer to the indicate parameters. Once the response is received ATT will destroy the ATT request object and call the callback with the pointer.

carlescufi commented:
@jhedberg do you want to take another look?

carlescufi commented on include/bluetooth/gatt.h:
```diff
@@ -1018,6 +1021,8 @@ struct bt_gatt_indicate_params {
 	/** Indicate Value callback */
 	bt_gatt_indicate_func_t func;
+	/** Host populated counter, this struct can be freed when equal to 0 */
+	int ref;
```
On second thought, let's be consistent with an already existing mechanism for this, the L2CAP destroy callback: 
https://github.com/zephyrproject-rtos/zephyr/blob/master/include/bluetooth/l2cap.h#L52

Perhaps `typedef void (*bt_gatt_indicate_params_destroy_t)(struct bt_gatt_indicate_params *params);`

One way to go about this would be to add the callback and then manage the reference count internally and expose only a `_ref` or `void *private` member in this structure.

JordanYates replied:
Where would this callback be added? To the `struct bt_gatt_indicate_params`?
If so, is this the accepted path forward?
My argument against it would be that we are adding another 4 bytes to the structure, when a usable callback is already run in the right places (I don't think this is this is the case for l2cap channels, but I'm not familiar with them).
The internal code then looks like.
```
    params->func(conn, params, err);
    params->_ref--;
    if (params->destroy && (params->_ref == 0)) {
        params->destroy(params);
    }
```

joerchan replied:
@JordanYates Yes, that would the accepted path forward.
To prevent increasing the size of the structure the _ref could be a uint16_t and placed at the end after len. The ref count would be maximum one per connection, so that should be enough. Could even be uint8_t I think.

Commit:
samples: Bluetooth: use indication destroy cb

Update the `peripheral` and `peripheral_ht` samples to use the
indication `destroy` callback to determine when the indication has
completed, instead of using the value callback of the first active
connection.

Signed-off-by: Jordan Yates <jordan.yates@data61.csiro.au>

JordanYates commented:
Implementation has been switched to the callback mechanism.
There is the potential for usage_faults in downstream code if the indication_params struct is not initialized as 0's (stack variable) and the user doesn't set the destroy callback, as the pointer will be random.
I don't believe this can be mitigated, but as this is already a stable API change I think its ok?
